### PR TITLE
[Workflow management UI changes 1] Implemented task chooser modal

### PR DESCRIPTION
**Pull request by kaedroho**
This adds a new task chooser modal. The modal allows the creation of new tasks or choosing an existing one.

This PR doesn't include:

 - Converting tasks to use plain Django forms
 - Admin listing fields for tasks
 - Fix for the checkbox widget styling

These will be included in future PRs.

Separated from: https://github.com/wagtail/wagtail/pull/6157

### CODE REVIEW

File: wagtail/admin/views/workflows.py [hunk 1]
```diff
@@ -62,7 +70,13 @@ def get_edit_handler(self):
         return self.edit_handler
 
     def get_form_class(self):
-        return self.get_edit_handler().get_form_class()
+        form_class = self.get_edit_handler().get_form_class()
+
+        # TODO Unhack
```
Comment (kaedroho) [inline wagtail/admin/views/workflows.py:75]: I've unhacked this in my other PR

File: wagtail/admin/views/workflows.py [hunk 2]
```diff
@@ -99,7 +113,13 @@ def get_edit_handler(self):
...
```
Comment (kaedroho) [inline wagtail/admin/views/workflows.py:118]: Ditto

File: wagtail/admin/templates/wagtailadmin/workflows/task_chooser/includes/results.html [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    {% endif %}
+
+    <table class="listing">
+        <thead>
+            <tr class="table-headers">
+                <th>
+                    {% trans "Name" %}
+                </th>
+                {% if searchform.specific_task_model_selected %}
+                    {# TODO #}
```
Comment (kaedroho) [inline wagtail/admin/templates/wagtailadmin/workflows/task_chooser/includes/results.html:22]: Going to leave this for a separate PR. We need to display some custom fields here and on the task index

File: wagtail/admin/forms/workflows.py [hunk 4]
```diff
@@ -1,9 +1,77 @@
...
+from django.utils.translation import ugettext as _
+from django.utils.translation import ugettext_lazy as __
 
 from wagtail.admin import widgets
-from wagtail.core.models import Page, Workflow, WorkflowPage
+from wagtail.core.models import Page, Task, Workflow, WorkflowPage
+
+
+def model_name(model):
+    return model._meta.app_label + '.' + model.__name__
```
Comment (zerolab) [inline wagtail/admin/forms/workflows.py:12]: could this live in `core.utils`? especially that it contains `resolve_model_string` which is the opposite of this, in a way

File: wagtail/admin/forms/workflows.py [hunk 5]
```diff
@@ -1,9 +1,77 @@
...
+                        for model, verbose_name in task_type_choices
+                    ]
+                ),
+                required=False
+            )
+
+        # Save a mapping of task_type values back to the model that we can reference later
+        self.task_type_choices = {
+            model_name(model): model
+            for model, _ in task_type_choices
```
Comment (zerolab) [inline wagtail/admin/forms/workflows.py:43]: can you rework this in light of @gasman's comment - https://github.com/wagtail/wagtail/pull/6163#issuecomment-647579099:

> for example, using _ as the return value of get_or_create is liable to clash with the _ alias commonly used for gettext

File: wagtail/admin/templates/wagtailadmin/workflows/task_chooser/chooser.html [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+    <section id="existing" class="nice-padding{% if not can_create %} active{% endif %}">
+        <form class="task-search search-bar" action="{% url 'wagtailadmin_workflows:task_chooser' %}" method="GET" novalidate>
+            <ul class="fields">
+                {% for field in searchform %}
+                    {% include "wagtailadmin/shared/field_as_li.html" with field=field %}
+                {% endfor %}
+            </ul>
+        </form>
+        <div id="search-results" class="listing tasks">
+            {% include "./includes/results.html" %}
```
Comment (zerolab) [inline wagtail/admin/templates/wagtailadmin/workflows/task_chooser/chooser.html:35]: All other chooser templates use full paths, let's be consistent. There are relative includes above too

File: wagtail/admin/templates/wagtailadmin/workflows/task_chooser/includes/results.html [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+            </tr>
+        </thead>
+        <tbody>
+            {% for task in tasks %}
+                <tr>
+                    <td class="title">
+                        <div class="title-wrapper"><a href="{% url 'wagtailadmin_workflows:task_chosen' task.id %}" class="task-choice">{{ task.name }}</a></div>
+                    </td>
+                    {% if searchform.specific_task_model_selected %}
+                        {# TODO #}
```
Comment (zerolab) [inline wagtail/admin/templates/wagtailadmin/workflows/task_chooser/includes/results.html:37]: Is this done in any of the follow-up PRs?
Can't see it in #6171 or #6172

File: wagtail/admin/views/workflows.py [hunk 8]
```diff
@@ -452,3 +472,146 @@ def enable_task(request, pk):
...
         return redirect('wagtailadmin_workflows:edit_task', task.id)
+
+
+def get_chooser_context():
+    """construct context variables needed by the chooser JS"""
+    return {
+        'step': 'chooser',
+        'error_label': _("Server Error"),
+        'error_message': _("Report this error to your webmaster with the following information:"),
+        'tag_autocomplete_url': reverse('wagtailadmin_tag_autocomplete'),
```
Comment (zerolab) [inline wagtail/admin/views/workflows.py:483]: This is not used anywhere, is it needed?

File: wagtail/admin/views/workflows.py [hunk 9]
```diff
@@ -452,3 +472,146 @@ def enable_task(request, pk):
...
+        edit_handler = create_model.get_edit_handler()
+        edit_handler = edit_handler.bind_to(request=request)
+        createform_class = edit_handler.get_form_class()
+    else:
+        createform_class = None
+
+    # TODO
+    #  # allow hooks to modify the queryset
+    # for hook in hooks.get_hooks('construct_document_chooser_queryset'):
+    #     documents = hook(documents, request)
```
Comment (zerolab) [inline wagtail/admin/views/workflows.py:542]: Looks like a remnant from copying the document_chooser method. Would be good to cleanup

File: wagtail/admin/templates/wagtailadmin/workflows/task_chooser/includes/results.html (hunk 7 above)
Comment (kaedroho) [reply wagtail/admin/templates/wagtailadmin/workflows/task_chooser/includes/results.html:37]: Nope, this hasn't been implemented yet

**Comment by kaedroho** [comment]
Thanks for the review @zerolab. I've addressed your feedback in: https://github.com/wagtail/wagtail/pull/6169/commits/e2359c914f23bbfaa1446f42773452b652c5b661

File: wagtail/admin/templates/wagtailadmin/workflows/task_chooser/includes/results.html (hunk 7 above)
Comment (zerolab) [reply wagtail/admin/templates/wagtailadmin/workflows/task_chooser/includes/results.html:37]: 👍
